Remove 5 destination limit, change activation cost, dialog cleanup

wispborne opened the pull request.

wispborne committed:
Set up IntelliJ project, converted to Kotlin

No changes made to logic.

wispborne committed:
Minor code cleanup

wispborne committed:
Changed cost of activation to 250 heavy machinery, 500 metal, and 1 gamma core

Updated text to show what user has in cargo.

wispborne committed:
Updating mod info

wispborne committed:
Cleaned up dialog a little

Trying to fly through non-active gate no longer closes dialog.
Destinations now use the star names, rather than the system id.
You can now reconsider jumping through the gate.

wispborne committed:
Cleaned up some more code

wispborne committed:
All activated gates are now available for travel

wispborne committed:
Fixed bug where trying to warp while gate was inactive would show the system id as an option

wispborne committed:
Fixing more instances of unnecessary precision in text.

wispborne committed:
Fixed fleet trying to return to the location of the gate they jumped from

wispborne committed:
Settled on "jump" as verb, updated readme, reworded activation words

wispborne committed:
Updated wording and attribution in readme

wispborne committed:
Changed project module name to Active-Gates

wispborne committed:
Clarify that it is working with 0.9a

wispborne committed:
Uncomment modThreadId

wispborne committed:
Set masterVersionFile back to toast's repo

wispborne committed:
Update authors to be less weird

ken-c commented on jars/src/org/toast/activegates/FlyThroughGate.kt:
```diff
@@ -0,0 +1,71 @@
+        val newVect = gates[0].location
+        playerFleet.setLocation(newVect.x, newVect.y)
+        playerFleet.clearAssignments()
+        playerFleet.setMoveDestination(playerFleet.location.x, playerFleet.location.y)
```
what was the intent of this line?

wispborne replied:
Whenever I'd jump through a gate, my fleet would start flying to the point in the new system that the gate we just jumped from had occupied. Like if gate A in system A is located at point A, and you jump to system B, my fleet would start flying to the location of point A but in system B.

I didn't spend much time making sure that I wasn't doing anything dumb, but adding this line of code fixed it.

ken-c commented on jars/src/org/toast/activegates/ActivateGate.kt:
```diff
@@ -0,0 +1,38 @@
+                    GateCommandPlugin.payActivationCost()
+                    gate.addTag(GateCommandPlugin.TAG_GATE_ACTIVATED)
+                    gate.memory
+                    textPanel.addParagraph("Your crew offload the resources and, almost reverently, a single gamma core." +
```
if the number of cores is changed, this text will also need to change

wispborne replied:
Yep. I don't like it but want to prioritize UX over maintainability - I think the best solution might be to move all visible strings to a single class so it's less likely that one would get missed if something changes.

ken-c commented on jars/src/org/toast/activegates/AddGateData.kt:
```diff
@@ -0,0 +1,37 @@
+import com.fs.starfarer.api.util.Misc
+
+class AddGateData : BaseCommandPlugin() {
+    override fun execute(ruleId: String, dialog: InteractionDialogAPI?,
```
since i know nothing about kotlin, what is the significance of the question mark?

wispborne replied:
Marks a type as nullable. Kotlin will enforce this both at compile time and runtime.

```kotlin
val nullableStr: String? = null // fine
val nonNullStr: String = null // compiler exception!
```

Since java doesn't have nullable types, when interopting with java you kinda just tell kotlin what to expect...and if you're wrong (eg in the example above, if we received a `null` `ruleId` from some java code), it'll explode at runtime.

Overall, a really nice system and one of the best arguments to use kotlin.

ken-c commented on jars/src/org/toast/activegates/AddGateData.kt:
```diff
@@ -0,0 +1,37 @@
+            textPanel.addParagraph("All activated gates:")
+
+            for (gate in GateCommandPlugin.getGateMap(GateFilter.Active)) {
+                textPanel.addParagraph("${gate.systemName} at ${gate.distanceFromPlayer}")
```
suggest we settle on a single string convention; either line 25 or line 31, but not both

wispborne replied:
Ha. Good point. I prefer the string template over concatenation, but didn't go through and change all of them.

I should note at this point that I didn't rewrite all of the code by hand. I used the built-in Java to Kotlin converter, then did sporadic refactoring. So, most of these inconsistencies are not an intentional choice I made to be inconsistent, but rather a clash of converted java code I didn't change and code I did change.

ken-c commented on jars/src/org/toast/activegates/GateCommandPlugin.kt:
```diff
@@ -0,0 +1,93 @@
+
+    private fun getCommodityCost(commodity: String): Float = activationCost[commodity] ?: 0F
+
+    const val TAG_GATE_ACTIVATED = "gate_activated"
```
maybe we should move away from "activate" and "activated" to "bring online" and "online", following the phrasing for comm relays, nav beacons, and sensor arrays.

wispborne replied:
Perhaps, but they _are_ referred to as "Inactive" in the game itself. I was hoping to be able to modify that at some point to read "Corvus Gate (Active)".

![screenshot006](https://user-images.githubusercontent.com/1343347/50752209-49051e80-121b-11e9-8086-9d6d49ba4291.png)

ken-c reviewed (commented):
made some random comments; most are minor, but i'm curious as to what broke without setMoveDestination

wispborne commented:
Thanks for the good comments. I'll have to resolve them another time...it's definitely past when I should be asleep.

wispborne committed:
Moved strings to a single location. More code comments.

Removed now-unnecessary handling of player jumping to their current system, since the current system is no longer an option.

wispborne committed:
Merge remote-tracking branch 'davidwhitman/wisp' into wisp

wispborne commented:
Alright, I've cleaned up the strings to be consistently declared and moved them to a single file. I also added a few code comments, although of course it could always use more.

ken-c commented:
thanks for updating!  the changes look good.
let's merge this, readjust the gate activation costs as per the feedback on the forum thread, and then i think we're ready for release.

wispborne commented:
Any thoughts on the quantity of transplutonics and volatiles? 

I'm thinking 100 of each. I don't see gates as unbalancing the game too incredibly, but rather reducing tedium. After all, you need to visit the location in order to activate the gate. 

And with colonies as they are, getting a large amount of money isn't much of a challenge, so abusing gates to quickly complete missions doesn't strike me as gamebreaking. Already, many players are saying that bounty missions usually make net zero profit because of the cost in supplies of getting to the fight + repairs, so if gates make bounties a viable source of income, that seems like a good thing for balance.

wispborne commented:
How does this look?

![2019-01-10_20-20-56_java](https://user-images.githubusercontent.com/1343347/51007391-3fc6cb00-1515-11e9-98fc-3c420dd7da86.png)

wispborne committed:
Added transplutonics and volatiles as required materials to activate a gate

wispborne commented on jars/src/org/toast/activegates/ActiveGatesStrings.kt:
```diff
@@ -0,0 +1,45 @@
+    fun notEnoughFuel(fuelCostOfJump: Int): String =
+            "Unfortunately, your fleet lacks the $fuelCostOfJump fuel necessary to use the gate."
+
+    val activationCost: String
```
Replaced the hardcoded description of required commodities with an auto-generated list.

ken-c commented:
Since the gates are activated on a one-time basis, i was thinking to have the cost be on the order of 2000 metals, 250-500 heavy machinery, 250 transplutonics, and 125-250 volatiles.  At least it provides an excuse to use the otherwise filler commodities that you run across, and means you need to have some reasonable cargo capacity.

i see whatever values we choose as temporary, to be readjusted after 0.9.1 is released.
and since they would be in settings.json, users can change them as they see fit.

ken-c commented:
Just looked at your last commit -- the framework looks good.  What do you think about the numbers I proposed?

wispborne commented:
I agree that the values are temporary. I had some ideas for more integration with the lore and a small quest needed, then you may activate a limited number of gates and that's it.

I'll leave the numbers decision up to you. 

Personally, I'll play using the numbers I chose. I started a new game a few days ago and rushed my first colony, but it's way out on the edge of the sector, so no way did I want to keep flying out there every time the pirates rang, and I was able to scrounge together the resources required and haul them out to my outpost. So, right now, they're about tuned to when the player is able to found their first colony. 

Another thought for a future version would be scaling up the costs dramatically based on the number of already activated gates (maybe the first two gates just require a gamma core but then it quickly scales up to multiple alpha cores because of the exponentially increasing complexity of maintaining the connection between more and more gates).

I'll make the changes to the numbers now.

wispborne committed:
Greatly increasing the resource cost to activate a gate

wispborne committed:
Activation materials no longer listed if their quantity is zero

wispborne committed:
Removing special mention of gamma core from strings.

It was only going to cause problems later.

ken-c commented:
last set of commits look good -- please merge when ready (i think you are able to do so since you are a collaborator; if not, then i need to investigate why you can't).
